doc/modules/array_api.rst

```diff
@@ -0,0 +1,73 @@
+.. _array_api:
+
+================================
+Array API support (experimental)
+================================
+
+.. currentmodule:: sklearn
+
+The `Array API <https://data-apis.org/array-api/latest/>`_ specification defines
+a standard API for all array manipulation libraries with a NumPy-like API.
+
+Some scikit-learn estimators that primarily rely on NumPy (as opposed to using
+Cython) to implement the algorithmic logic of their `fit`, `predict` or
+`transform` methods can be configured to accept any Array API compatible input
+datastructures and automatically dispatch operations to the underlying namespace
+instead of relying on NumPy.
+
+At this stage, this support is **considered experimental** and must be enabled
+explicitly as explained in the following.
+
+Example usage
+=============
+
+Here is an example code snippet to demonstrate how to use `CuPy
+<https://cupy.dev/>`_ to run
+:class:`~sklearn.discriminant_analysis.LinearDiscriminantAnalysis` on a GPU::
+
+    >>> from sklearn.datasets import make_classification
+    >>> from sklearn import config_context
+    >>> from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+    >>> import cupy.array_api as xp
+
+    >>> X_np, y_np = make_classification(random_state=0)
+    >>> X_cu = xp.asarray(X_np)
+    >>> y_cu = xp.asarray(y_np)
+    >>> X_cu.device
+    <CUDA Device 0>
+
+    >>> with config_context(array_api_dispatch=True):
+    ...     lda = LinearDiscriminantAnalysis()
+    ...     X_trans = lda.fit_transform(X_cu, y_cu)
+    >>> X_trans.device
+    <CUDA Device 0>
+
+After the model is trained, fitted attributes that are arrays will also be
+from the same Array API namespace as the training data. For example, if CuPy's
+Array API namespace was used for training, then fitted attributes will be on the
+GPU. Moving the fitted attributes to the CPU depends on the array library. For
+example, if you use `cupy.array_api` then the transfer can be done as follows::
+
+.. code-block:: python
+
+    >>> for key, attribute in vars(lda).items():
+    ...     if not hasattr(attribute, "__array_namespace__"):
+    ...         # Does not implement the Array API specification
+    ...         continue
+    ...     # Convert `cupy.array_api` array into a NumPy ndarray
+    ...     setattr(lda, key, attribute._array.get())
+    >>> # transform on a CPU
+    >>> X_trans = lda.transform(X_np)
+    >>> type(X_trans)
+    <class 'numpy.ndarray'>
```

Comment:
We might want to provide an experimental helper, e.g. `_convert_to_xp(model, target_namespace)` to achieve this instead of putting an ugly code snippet in the doc. Here `target_namespace` can be an `xp` module or a namespace path (`str`).

Reply:
Actually here we would rather need `_convert_to_np` or something similar.

Reply:
The Array API specification pushes for [dlpack](https://data-apis.org/array-api/latest/design_topics/data_interchange.html) to do data exchange between different Array API implementation.

I think it mostly works if the array is on the same device. For arrays on different devices, there is no public API that "just works".

With that in mind, I added a `_convert_estimator_to_ndarray` that converts from `cupy.array_api` or `numpy.array_api` into a `numpy.ndarray`.

Reply:
A bit of context: there was a common understanding that automatic/implicit transfer of data between different devices is not a good idea and leads to hard to detect performance issues. It should be an explicit action by the user. That is why there is nothing that "just works".

Reply:
Is there an API for a user to explicitly convert? For example, I hoped `from_dlpack` worked:

```python
import cupy.array_api as cu_xp
import numpy.array_api as np_xp

x = cu_xp.asarray([1, 2, 3.4])

# RuntimeError: Unsupported device in DLTensor.
x_np = np_xp.from_dlpack(x)
```

Currently, we need to use `cupy.ndarray` API to do the device transfer first, and then NumPy can convert:

```python
x_np = np_xp.from_dlpack(x_cu._array.get())
```

Reply:
No, there isn't such an API, and no plan for it. Certainly shouldn't be mixed with DLPack I think, which is explicitly a zero-copy protocol (just like the buffer protocol, `__array_interface__`, `__cuda_array_interface__`, etc.). There are only library-specific methods for device transfers. xref https://github.com/pytorch/pytorch/issues/36560 also for a relevant discussion in PyTorch, where a `force=` keyword was requested (makes _some_ sense, but not implemented).

Before saying more, I should probably look at the code to understand why you want this, rather than leaving it to the user.

Reply:
The usecase is:

- fit a scikit-learn estimator on a GPU machine: the internal state are CuPy arrays backed by a GPU device managed by a CUDA runtime;
- convert the estimator object to use numpy arrays instead;
- pickle the trained model to a file, model store, docker container image, whatever;
- load the pickle in a Python process on a machine without CUDA to compute predictions on new data using only the CPU and no extra dependencies beyond numpy and scipy.

Reply:
That makes sense. In general, it's not supportable at least in the short term to have forcing (regular copy, device transfer, detach from autograd graph, etc.) conversions between any two array types from different libraries. Some libraries currently don't even offer this for "own kind to numpy", and a numpy-specific idiom won't be very attractive for maintainers of other libraries.

At the moment, the right thing to use is `cupy.asnumpy()` or `cupy.ndarray.get()` (see https://docs.cupy.dev/en/stable/reference/ndarray.html#conversion-to-from-numpy-arrays). You know you want a "to numpy" call, which is an easier ask than a "any to any other" generic API. There may be appetite to align the names of `cupy.asnumpy()`/`torch.Tensor.numpy()`/`tf.Tensor.numpy()` names, that seems like a logical next step. For now I'd say you want to special-case the libraries in some helper routine.

Reply:
If an array library implementation provides a convertor, it would be nice if scikit-learn provided a mapper over attributes of trained estimator to apply that function. It would be `sklearn.utils.convert(est, convertor_func)`. This would create a copy of the estimator where trained array attributes were converted using the given function.

This would allow to train on GPU and predict on CPU, for example.

Reply:
Indeed we could have a private `_convert_attribute(est, converter_func)` where `converter_func` could be a lambda expression and use that in the doc.

Reply:
I am happy with having a `attribute_converter_func` in `_convert_array_attributes` for users to provided library specific conversions.

I would go as far as making the converter function a requirement. This way we do not need to maintain library specific code in scikit-learn for array conversions.

Reply:
That would be great.